drivers/sx126x/sx126x_radio_hal.c

```diff
@@ -0,0 +1,668 @@
+    }
+
+    #if IS_USED(MODULE_SX126X_STM32WL)
+        NVIC_EnableIRQ(SUBGHZ_Radio_IRQn);
+    #endif  
+}
+}
+
+                        /*RADIO HAL FUNCTIONS*/
+
+static int _write(ieee802154_dev_t *hal, const iolist_t *iolist){
+
+    sx126x_t *dev = hal->priv;
+    (void)dev;
+    size_t pos = 0;
+    /* Full buffer used for Tx */
+    sx126x_set_buffer_base_address(dev, 0x80, 0x00);
+    /* Write payload buffer */
+    for (const iolist_t *iol = iolist; iol; iol = iol->iol_next) {
+        if (iol->iol_len > 0) {
+            sx126x_write_buffer(dev, pos + 0x80, iol->iol_base, iol->iol_len);
+            DEBUG("[sx126x] netdev: send: wrote data to payload buffer.\n");
+            pos += iol->iol_len;
+        }
+    }
+
+    sx126x_set_lora_payload_length(dev, pos);
+    DEBUG("[sx126x] writing (size: %d).\n", (pos));
+    return 0;
+}
+
+static int _request_op(ieee802154_dev_t *hal, ieee802154_hal_op_t op, void *ctx){
+    sx126x_t *dev = hal->priv;
+    (void)dev;
+    (void)ctx;
+    switch (op) {
+        case IEEE802154_HAL_OP_TRANSMIT:
+        dev->pending = false;
+        ieee802154_radio_cca(hal);
+        _set_state(dev, STATE_TX);
+
+        break;
+        case IEEE802154_HAL_OP_SET_RX:
+
+        _set_state(dev, STATE_RX);
+
+        break;
+
+        case IEEE802154_HAL_OP_SET_IDLE:
+
+        _set_state(dev, STATE_IDLE);
+
+        break;
+
+        case IEEE802154_HAL_OP_CCA:
+            DEBUG("[sx126x] netdev: HAL_OP_CCA (CAD Detection state)\n");
+            dev->cad_detected = 0;
+            _set_state(dev, STATE_IDLE);
+            sx126x_set_cad(dev);
+
+        break;
+
+        default:
+        assert(false);
+        break;
+
+    }
+    return 0;
+}
+
+static int _confirm_op(ieee802154_dev_t *hal, ieee802154_hal_op_t op, void *ctx){
+    sx126x_t *dev = hal->priv;
+    (void)op;
+    (void)ctx;
+    ieee802154_tx_info_t *info = ctx;
+    bool eagain = false;
+    sx126x_state_t state;
+    _get_state(dev, &state);
+
+    
+switch (op){
+    case IEEE802154_HAL_OP_TRANSMIT:
+       if (info) {
+            info->status = (dev->cad_detected) ? TX_STATUS_MEDIUM_BUSY : TX_STATUS_SUCCESS;
+        }
+
+        
+    break;
+
+    case IEEE802154_HAL_OP_SET_RX:
+        if(state!=STATE_RX) 
+            eagain = true;
+    break;
+
+    case IEEE802154_HAL_OP_SET_IDLE:
+        eagain = (state != STATE_IDLE);
+        break;
+
+    case IEEE802154_HAL_OP_CCA:
+            *((bool*) ctx) = !dev->cad_detected;
+    break;
+
+    default:
+        eagain = false;
+        break;
+    
+}
+
+    if (eagain) {
+       
+        return -EAGAIN;
+    }
+
+    return 0;
+}
+
+static int _len(ieee802154_dev_t *hal){
+    (void)hal;
+    sx126x_t *dev = hal->priv;
+    sx126x_rx_buffer_status_t rx_buffer_status;
+    sx126x_get_rx_buffer_status(dev, &rx_buffer_status);
+    return rx_buffer_status.pld_len_in_bytes;
+}
+
+static int _read(ieee802154_dev_t *hal, void *buf, size_t max_size, ieee802154_rx_info_t *info)
```

Comment:
Same here for the CRC: here we calculate the CRC and compare it with the one from the frame.

You can base on this snippet (please ignore the debug messages, and the code is a bit old)

```diff
@@ -450,21 +351,34 @@ static int _read(ieee802154_dev_t *hal, void *buf, size_t max_size, ieee802154_r
     (void)hal;
     (void)buf;
     (void)info;
+    uint16_t chksum = 0;
+    uint16_t exp_chksum;
 
+    DEBUG("[sx126x] _read\n");
     sx126x_t* dev = hal->priv;
 
         /* Getting information about last received packet */
-    netdev_lora_rx_info_t *packet_info = (void*)info;
+    //netdev_lora_rx_info_t *packet_info = (void*)info;
     sx126x_rx_buffer_status_t rx_buffer_status;
     sx126x_pkt_status_lora_t pkt_status;
     sx126x_get_rx_buffer_status(dev, &rx_buffer_status);
 
+    /* Size including CRC */
     dev->size = rx_buffer_status.pld_len_in_bytes;
+
     sx126x_get_lora_pkt_status(dev, &pkt_status);
+
+    /* TODO */
+    if (info) {
+        info->lqi = 255;
+        info->rssi = 100;
+    }
+    /*
     if (packet_info) {
         packet_info->snr = pkt_status.snr_pkt_in_db;
         packet_info->rssi = pkt_status.rssi_pkt_in_dbm;
     }
+    */
 
       /* Put PSDU to the output buffer */
     
@@ -472,14 +386,26 @@ static int _read(ieee802154_dev_t *hal, void *buf, size_t max_size, ieee802154_r
         return dev->size;
     }
 
-    if (dev->size > max_size) {
+    /* Exclude CRC */
+    if (dev->size > max_size + 2) {
         return -ENOBUFS;
     }
 
     if (dev->size < 3) {
         return -EBADMSG;
     }
-    sx126x_read_buffer(dev, rx_buffer_status.buffer_start_pointer, (uint8_t*)buf, dev->size);
+    /* Copy to buffer */
+    sx126x_read_buffer(dev, rx_buffer_status.buffer_start_pointer, (uint8_t*)buf, dev->size-2);
+    sx126x_read_buffer(dev, rx_buffer_status.buffer_start_pointer + dev->size-2, (uint8_t*)&exp_chksum, sizeof(exp_chksum));
+    chksum = ucrc16_calc_le(buf, dev->size-2,
+                        UCRC16_CCITT_POLY_LE, chksum);
+    chksum = byteorder_htols(chksum).u16;
+    /* Validate checksum */
+    if (chksum != exp_chksum) {
+        puts("CRC_F");
+        return -EBADMSG;
+    }
+
     DEBUG("[sx126x] first 3 bytes of received packet: %d %d %d\n", *(uint8_t*)buf, *((uint8_t*)buf+1), *((uint8_t*)buf+2));
     return dev->size;
 }
```